src/bucket.ts

```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
+   *
+   * @example
+   * const {Storage} = require('@google-cloud/storage');
+   * const storage = new Storage();
+   * const bucket = storage.bucket('albums');
+   *
+   * //-
+   * // Upload a directory from a local path.
+   * //-
+   * bucket.uploadDirectory('/local/path/myfolder', function(err, resp) {
+   *   // Your bucket now contains:
+   *   // - "myfolder/" directory (with the contents of `/local/path/myfolder/')
+   *
+   *   // `file` is an instance of a File object that refers to your new file.
+   * });
+   *
+   *
+   * //-
+   * // You can also have files gzip'd on the fly.
+   * //-
+   * bucket.uploadDirectory(path.join(__dirname, 'index', { gzip: true }, function(err, resp) {
+   *   // Your bucket now contains:
+   *   // - "index/" directory (automatically compressed with gzip)
+   *
+   *   // Downloading the file with `file.download` will automatically decode
+   * the
+   *   // file.
+   * });
+   *
+   *
+   * //-
+   * // If the callback is omitted, we'll return a Promise.
+   * //-
+   * bucket.uploadDirectory('local-directory-path').then(function(data) {
+   *   const resp = data[0];
+   * });
+   */
+
+  uploadDirectory(
+    directoryPathString: string,
+    optionsOrCallback?: UploadDirectoryOptions | UploadDirectoryCallback,
+    callback?: UploadDirectoryCallback
+  ): Promise<UploadDirectoryResponse> | void {
+    // tslint:disable-next-line:no-any
+    if ((global as any)['GCLOUD_SANDBOX_ENV']) {
+      return;
+    }
+
+    const options =
+      typeof optionsOrCallback === 'object' ? optionsOrCallback : {};
+    callback =
+      typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
+
+    let stat: fs.Stats | Error;
+    try {
+      stat = fs.lstatSync(directoryPathString);
+    } catch (error) {
+      stat = error;
+    }
+    if (stat instanceof Error || !stat.isDirectory()) {
+      throw new Error(`${directoryPathString} is an invalid directory`);
+    }
+
+    const fileList: string[] = [];
+    const pathDirName = path.dirname(directoryPathString);
+
+    getFileList(directoryPathString);
+
+    function getFileList(directory: string) {
+      const items = fs.readdirSync(directory);
+      items.forEach(item => {
+        const fullPath = path.join(directory, item);
+        const stat = fs.lstatSync(fullPath);
+        if (stat.isFile()) {
+          fileList.push(fullPath);
+        } else if (options.recurse && stat.isDirectory()) {
+          getFileList(fullPath);
+        }
+      });
+    }
+
+    const limit = pLimit(10);
+    Promise.all(
+      fileList.map(file =>
+        limit(() => {
+          let destination = path.relative(pathDirName, file);
+          if (process.platform === 'win32') {
+            destination = destination.replace(/\\/g, '/');
+          }
+          const uploadOptions = extend({}, {destination}, options);
+          return this.upload(file, uploadOptions).then(
+            () => ({fileName: uploadOptions.destination, status: 'success'}),
```

Comment:
I feel like we're definitely going to want to pass back the entire response body somehow

Reply:
Do you prefer something like
```ts
return this.upload(file, uploadOptions).then(
            (res) => ({fileName: uploadOptions.destination, status: res[0]}),
            (err: Error) => ({fileName: uploadOptions.destination, status: err})
          );
```
`File` object contains `metadata` property which is the `Metadata` part of the `bucket.upload` response. So essentially we will be returning entire `bucket.upload` response.
Please share you thoughts.

Reply:
Ah ok, that sounds good to me! Thanks for the break down